src/coprocessor/dag/expr/compare.rs

```diff
@@ -0,0 +1,294 @@
+                } as i64))
+            }
+        }
+    }
+
+    pub fn compare_duration(&self,
+                            ctx: &StatementContext,
+                            row: &[Datum],
+                            sig: ScalarFuncSig)
+                            -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_duration(ctx, row));
+        let rhs = try!(self.children[1].eval_duration(ctx, row));
+        match (lhs, rhs) {
+            (None, _) | (_, None) => {
+                if sig == ScalarFuncSig::NullEQDuration {
+                    Ok(Some(1))
+                } else {
+                    Ok(None)
+                }
+            }
+            (Some(lhs), Some(rhs)) => {
+                let ordering = lhs.cmp(&rhs);
+                Ok(Some(match sig {
+                    ScalarFuncSig::LTDuration => ordering == Ordering::Less,
+                    ScalarFuncSig::LEDuration => ordering != Ordering::Greater,
+                    ScalarFuncSig::GTDuration => ordering == Ordering::Greater,
+                    ScalarFuncSig::GEDuration => ordering != Ordering::Less,
+                    ScalarFuncSig::EQDuration => ordering == Ordering::Equal,
+                    ScalarFuncSig::NEDuration => ordering != Ordering::Equal,
+                    _ => unreachable!(),
+                } as i64))
+            }
+        }
+    }
+
+    pub fn compare_json(&self,
+                        ctx: &StatementContext,
+                        row: &[Datum],
+                        sig: ScalarFuncSig)
+                        -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_json(ctx, row));
+        let rhs = try!(self.children[1].eval_json(ctx, row));
+        match (lhs, rhs) {
+            (None, _) | (_, None) => {
+                if sig == ScalarFuncSig::NullEQJson {
+                    Ok(Some(1))
+                } else {
+                    Ok(None)
+                }
+            }
+            (Some(lhs), Some(rhs)) => {
+                let ordering = lhs.cmp(&rhs);
+                Ok(Some(match sig {
+                    ScalarFuncSig::LTJson => ordering == Ordering::Less,
+                    ScalarFuncSig::LEJson => ordering != Ordering::Greater,
+                    ScalarFuncSig::GTJson => ordering == Ordering::Greater,
+                    ScalarFuncSig::GEJson => ordering != Ordering::Less,
+                    ScalarFuncSig::EQJson => ordering == Ordering::Equal,
+                    ScalarFuncSig::NEJson => ordering != Ordering::Equal,
+                    _ => unreachable!(),
+                } as i64))
+            }
+        }
+    }
+}
+
+#[inline]
+fn cmp_i64_with_unsigned_flag(lhs: i64,
+                              lhs_unsigned: bool,
+                              rhs: i64,
+                              rhs_unsigned: bool)
+                              -> Ordering {
+    match (lhs_unsigned, rhs_unsigned) {
+        (true, true) | (false, false) => lhs.cmp(&rhs),
```

Comment:
This is wrong. For example, when comparing `u64::MAX` to `5u64`, this function will return `Ordering::Less`.